tests/unittests/client/test_runner.py

```diff
@@ -0,0 +1,207 @@
+
+
+def test_pending_idle_worker():
+    """No new trials can be generated but we have a pending trial so LazyWorkers is not raised."""
+    idle_timeout = 1
+    pop_time = 1
+    runner = new_runner(idle_timeout)
+
+    # Dummy pending that will prevent runner from
+    # raising LazyWorkers
+    runner.pending_trials[0] = None
+
+    def remove_pending():
+        time.sleep(pop_time)
+        runner.pending_trials = dict()
+
+    start = time.time()
+    thread = Thread(target=remove_pending)
+    thread.start()
+
+    with pytest.raises(LazyWorkers):
+        runner.run()
+
+    elapsed = time.time() - start
+
+    assert (
+        int(elapsed - (pop_time + idle_timeout)) == 0
+    ), "LazyWorkers was raised after pending_trials got emptied"
+
+
+def test_no_remaining_worker():
+    """Runner stops if we have not more trials to run"""
+    idle_timeout = 1
+    pop_time = 1
+    runner = new_runner(idle_timeout)
+
+    runner.pending_trials[0] = None
+
+    def no_more_trials():
+        time.sleep(pop_time)
+        runner.pending_trials = dict()
+        runner.trials = 2
+
+    start = time.time()
+    thread = Thread(target=no_more_trials)
+    thread.start()
+
+    # Lazy worker is not raised because we have executed
+    # the max number of trials on this worker
+    runner.run()
+
+    elapsed = time.time() - start
+
+    assert (
+        int(elapsed - pop_time) == 0
+    ), "Runner terminated gracefully once max trials was reached"
+
+
+def test_is_done_worker():
+    """Runner stops when the experiment is_done"""
+    idle_timeout = 1
+    pop_time = 1
+    runner = new_runner(idle_timeout)
+
+    runner.pending_trials[0] = None
+
+    def set_is_done():
+        time.sleep(pop_time)
+        runner.pending_trials = dict()
+        runner.client.is_done = True
+
+    start = time.time()
+    thread = Thread(target=set_is_done)
+    thread.start()
+
+    runner.run()
+
+    elapsed = time.time() - start
+
+    assert (
+        int(elapsed - pop_time) == 0
+    ), "Runner terminated gracefully once experiment is done"
+
+
+def test_should_sample():
+    """Should sample should return the number of trial we can sample"""
+
+    def make_runner(n_workers, max_trials_per_worker, pool_size=None):
+        if pool_size is None:
+            pool_size = n_workers
+
+        return Runner(
+            client=FakeClient(n_workers),
+            fct=function,
+            pool_size=pool_size,
+            idle_timeout=1,
+            max_broken=2,
+            max_trials_per_worker=max_trials_per_worker,
+            trial_arg=[],
+            on_error=None,
+        )
+
+    assert (
+        make_runner(5, 2).should_sample() == 2
+    ), "5 processes but only 2 trials allowed"
+
+    assert (
+        make_runner(2, 5).should_sample() == 2
+    ), "2 processes and 5 max trials allowed"
+
+    assert (
+        make_runner(5, 5, 2).should_sample() == 2
```

Comment:
```suggestion
        make_runner(5, 5, 2).should_sample() == 5
```
It should be 5 as explained here: https://github.com/Epistimio/orion/pull/684/files#r770814583